src/transform/src/analysis/monotonic.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Analysis that determines the logical monotonicity of an expression.
+//!
+//! Expressions are monotonic if their output collection does not contain
+//! any retractions.
```

Comment:
Nit: This starts by talking about logical monotonic., but then doesn't define it. The second sentence is only true if the expression is logically monotonic _and_ the input is physically monotonic.